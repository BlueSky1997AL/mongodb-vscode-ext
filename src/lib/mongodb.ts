'use strict';

import MongoDBMemoryServer from 'mongodb-memory-server';
import { homedir } from 'os';
import { existsSync, mkdirsSync } from 'fs-extra';
import * as vscode from 'vscode';

import statusBarItem from './status-bar-item';

class MongoDB {

  public mongoDBInstance = null;

  start(options: any = {}) {
    const { port, dbName, dbPath, storageEngine, instanceDebug, binaryVersion, downloadDir, platform, arch, binaryDebug, mongoDebug, autoStart } = options

    let finalDbPath = dbPath || `${homedir()}/.mongodb/data`;
    if (!existsSync(finalDbPath)) {
      mkdirsSync(finalDbPath);
    }

    const mongod = new MongoDBMemoryServer({
      instance: {
        port: parseInt(port) || 27017,
        dbName: dbName || 'db',
        dbPath: finalDbPath,
        storageEngine,
        debug: instanceDebug || false
      },
      binary: {
        version: binaryVersion,
        downloadDir: downloadDir || `${homedir()}/.mongodb/binaries`,
        platform,
        arch,
        debug: binaryDebug
      },
      debug: mongoDebug,
      autoStart
    });

    mongod.start();

    this.mongoDBInstance = mongod;

    statusBarItem.update({
      text: '$(database) MongoDB',
      command: 'extension.showMenu',
      color: '#69b241',
      tooltip: 'MongoDB 数据库已启动, 单击以查看选项'
    })
  }

  stop() {
    this.mongoDBInstance.stop();
    this.mongoDBInstance = null;
    statusBarItem.update({
      text: '$(database) MongoDB',
      command: 'extension.showMenu',
      tooltip: 'MongoDB 数据库未启动，单击启动'
    });
  }

  get instance() {
    return this.mongoDBInstance;
  }

}

export default new MongoDB();
