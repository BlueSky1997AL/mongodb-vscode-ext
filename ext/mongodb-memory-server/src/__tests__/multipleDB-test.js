/* @flow */

import { MongoClient } from 'mongodb';
import MongoDBMemoryServer from '../MongoMemoryServer';

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

let con1;
let con2;
let db1;
let db2;
let mongoServer1;
let mongoServer2;
beforeAll(async () => {
  mongoServer1 = new MongoDBMemoryServer();
  const mongoUri = await mongoServer1.getConnectionString();
  con1 = await MongoClient.connect(mongoUri);
  db1 = con1.db(await mongoServer1.getDbName());

  mongoServer2 = new MongoDBMemoryServer();
  const mongoUri2 = await mongoServer2.getConnectionString();
  con2 = await MongoClient.connect(mongoUri2);
  db2 = con2.db(await mongoServer1.getDbName());
});

afterAll(() => {
  con1.close();
  con2.close();
  mongoServer1.stop();
  mongoServer2.stop();
});

describe('Multiple mongoServers', () => {
  it('should start several servers', async () => {
    expect(db1).toBeDefined();
    const col1 = db1.collection('test');
    const result1 = await col1.insert([{ a: 1 }, { b: 1 }]);
    expect(result1.result).toMatchSnapshot();
    expect(await col1.count({})).toBe(2);

    expect(db2).toBeDefined();
    const col2 = db2.collection('test');
    const result2 = await col2.insert([{ a: 2 }, { b: 2 }]);
    expect(result2.result).toMatchSnapshot();
    expect(await col2.count({})).toBe(2);
  });

  it('should have different uri', async () => {
    const uri1 = await mongoServer1.getConnectionString();
    const uri2 = await mongoServer2.getConnectionString();

    expect(uri1).not.toEqual(uri2);
  });
});
