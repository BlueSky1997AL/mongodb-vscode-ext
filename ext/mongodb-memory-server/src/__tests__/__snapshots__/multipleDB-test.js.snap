// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple mongoServers should start several servers 1`] = `
Object {
  "n": 2,
  "ok": 1,
}
`;

exports[`Multiple mongoServers should start several servers 2`] = `
Object {
  "n": 2,
  "ok": 1,
}
`;
